### ISSUE TITLE
Vim9: cannot use boolean number where bool is expected

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7644]
**Describe the bug**

In Vim9 script, sometimes, we cannot use a boolean number where a bool is expected in a `:def` function.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9
        def Func()
            var b: bool
            var s = '0'
            b = s->str2nr()
        enddef
        defcompile
    EOF
    )

`E1012` is raised:

    E1012: Type mismatch; expected bool but got number

**Expected behavior**

No error is raised, because `s->str2nr()` is the number `0`, and since [8.2.1641](https://github.com/vim/vim/releases/tag/v8.2.1641), it is allowed to use the numbers 0 or 1 where a bool is expected.

**Environment**

 - Vim version: 8.2 Included patches: 1-2318
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(363)

**Additional context**

The exact same code doesn't raise any error at the script level:
```vim
vim9
var b: bool
var s = '0'
b = s->str2nr()
```
---

Here are 2 other similar snippets which unexpectedly raise an error:
```vim
vim9
def Func(): bool
    var s = '0'
    return s->str2nr()
enddef
defcompile
```
    E1012: Type mismatch; expected bool but got number
```vim
vim9
def Func(): bool
    var n: number = 0
    return n
enddef
defcompile
```
    E1012: Type mismatch; expected bool but got number

The last one is really weird, because the error disappears if we don't specify the type of `n`:
```vim
vim9
def Func(): bool
    var n = 0
    return n
enddef
defcompile
```
    ✔

But that should not matter, because the type `number` is inferred from the value `0` anyway.

**Comment by brammool** [comment]
str2nr() returns a number, thus assigning it to a bool generally is wrong.  That the value happens to be OK for a bool can only be found at runtime, and in this case it's not really the intended use of this feature.
You can use !!str2nr() to convert the number to a bool.

At the script level everything is done at runtime, thus that's slightly different.  You won't get a compilation error, obviously.

When using "var nr = 0" Vim knows that the zero might be used as a bool, a special flag is set on the type of "nr" to allow it to be used as a bool.  This was added for special cases.  It might be unexpected, I don't think we can turn it into an error without breaking something.

**Comment by lacygoill** [comment]
This snippet doesn't raise any error:
```vim
vim9script
def Func()
    var n = 0
    n = 3
    if n
        echo 'true'
    else
        echo 'false'
    endif
enddef
Func()
```
    true

Is it working as intended?

---

This one doesn't work:
```vim
vim9script
def Func()
    var n = 0
    n = 3
    echo n ? 'true' : 'false'
enddef
Func()
```
    E1023: Using a Number as a Bool: 3

Why does the special flag which allows `n` to be used as a bool work in an `if` test, but not when passed as an operand to `?:`?

**Comment by lacygoill** [comment]
> Is it working as intended?

Fixed by 8.2.2755.

---

Something else looks weird:
```vim
vim9script
def Func()
    var n: number = 1
    if n
        echo 'true'
    endif
enddef
Func()
```
    true

Shouldn't this snippet raise an error (at compile time)?

We've explicitly specified that the type of `n` was a number, so Vim shouldn't set the special flag, right?  Exactly like here, Vim doesn't set it:
```vim
vim9script
def Func()
    var n: number = 1
    var b: bool = n
enddef
defcompile
```
    E1012: Type mismatch; expected bool but got number

Why is it ok for `n` – whose type is explicitly `number` (no type inferrence) – to be used as a bool in an `if` test, but not in an assignment to a variable of type `bool`?

**Comment by brammool** [comment]
> Something else looks weird:
> ```vim
> vim9script
> def Func()
>     var n: number = 1
>     if n
>         echo 'true'
>     endif
> enddef
> Func()
> ```
>     true
> 
> Shouldn't this snippet raise an error (at compile time)?

No, the number zero and one can be used for false and true in a
condition.  I don't think it makes sense to be stricter in some places,
since we can't do that everywhere (e.g. builtin functions returning
these values).
 
> We've explicitly specified that the type of `n` was a number, so Vim shouldn't set the special flag, right?  Exactly like here, Vim doesn't set it:
> ```vim
> vim9script
> def Func()
>     var n: number = 1
>     var b: bool = n
> enddef
> defcompile
> ```
>     E1012: Type mismatch; expected bool but got number
> 
> Why is it ok for `n` – whose type is explicitly `number` (no type
> inferrence) – to be used as a bool in an `if` test, but not in an
> assignment to a variable of type `bool`?

When an expression is used as a condition we allow for zero and one.
The value itself is still a number.

-- 
From "know your smileys":
 :-D	Big smile

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
